### PR TITLE
Add Prometheus metrics reader factory and config

### PR DESCRIPTION
**Pull request by albertteoh**
## Which problem is this PR solving?
- Relates to #2954.

## Short description of the changes
- Add factories and config for Prometheus metrics reader.

### CODE REVIEW

File: pkg/prometheus/config/config.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import "time"
+
+// Configuration describes the options to customize the storage behavior
```
Comment (jpkrohling) [inline pkg/prometheus/config/config.go:19]:
```suggestion
// Configuration describes the options to customize the storage behavior.
```

File: plugin/metrics/factory.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+)
+
+const (
+	prometheusStorageType = "prometheus"
+)
+
+// AllStorageTypes defines all available storage backends.
+var AllStorageTypes = []string{prometheusStorageType}
+
+// Factory implements storage.Factory interface as a meta-factory for storage components.
```
Comment (jpkrohling) [inline plugin/metrics/factory.go:37]: Can you add a type-assertion, to make sure that this struct always implements the interface?

File: plugin/metrics/factory.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	}
+	return f, nil
+}
+
+func (f *Factory) getFactoryOfType(factoryType string) (storage.MetricsFactory, error) {
+	switch factoryType {
+	case prometheusStorageType:
+		return prometheus.NewFactory(), nil
+	}
+	return nil, fmt.Errorf("unknown metrics type %s. Valid types are %v", factoryType, AllStorageTypes)
```
Comment (jpkrohling) [inline plugin/metrics/factory.go:65]:
```suggestion
	return nil, fmt.Errorf("unknown metrics type %q. Valid types are %v", factoryType, AllStorageTypes)
```

File: plugin/metrics/factory.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+		factory.Initialize(logger)
+	}
+	return nil
+}
+
+// CreateMetricsReader implements storage.MetricsFactory.
+func (f *Factory) CreateMetricsReader() (metricsstore.Reader, error) {
+	factory, ok := f.factories[f.MetricsStorageType]
+	if !ok {
+		return nil, fmt.Errorf("no %s backend registered for metrics store", f.MetricsStorageType)
```
Comment (jpkrohling) [inline plugin/metrics/factory.go:80]:
```suggestion
		return nil, fmt.Errorf("no %q backend registered for metrics store", f.MetricsStorageType)
```

File: plugin/metrics/factory_test.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+	"github.com/spf13/viper"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+
+	"github.com/jaegertracing/jaeger/storage"
+	"github.com/jaegertracing/jaeger/storage/mocks"
+)
+
+var _ storage.MetricsFactory = new(Factory)
```
Comment (jpkrohling) [inline plugin/metrics/factory_test.go:30]: I see that you did the type assertion here. Forget my previous comment on this.

File: plugin/metrics/prometheus/options.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	"strings"
+	"time"
+
+	"github.com/spf13/viper"
+
+	"github.com/jaegertracing/jaeger/pkg/prometheus/config"
+)
+
+const (
+	suffixHostPort       = ".host-port"
```
Comment (jpkrohling) [inline plugin/metrics/prometheus/options.go:28]: Please, tell me that the next change is about adding TLS support to the communication with Prometheus :-)

**Review by jpkrohling — APPROVED** [review]

**Review by jpkrohling — APPROVED** [review]

**Comment by albertteoh** [comment]
Thanks @jpkrohling!

**Pull request merged**